contrib: Magic, items systems for 'turnbattle'

FlutterSprite opened the pull request:
#### Brief overview of PR changes/additions
Adds another version of the turn based combat system that includes a system for casting magic spells or using similar special abilities - spells are defined in a dictionary that determines their cost, valid spell targets, and the circumstances under which the spell can be cast. Spell names are matched to functions, and the spell definitions can pass kwargs to them, meaning the same function can be re-used to make multiple spells.

This also adds another module for usable items and status effects - items work similarly to spells, executing arbitrary functions, and have some options for limited uses, consuming spent items, and replacing spent items with another object as 'residue' (such as a glass bottle being left behind after using a potion). This module also includes conditions, temporary effects on characters that can be beneficial or harmful, with set durations that count down each turn or every 30 seconds when out of combat.

#### Motivation for adding to Evennia
Magic systems are a fairly common element of MUDs, so a ready-to-go system for spellcasting is something that may be useful to a lot of developers - it also serves as a good example of a few code techniques that may not be immediately obvious to Python beginners, like storing references to functions and passing arbitrary kwargs.

Similarly, items and status effects are a big part of RPGs in general and MUDs in particular - the items system is very flexible, and can be used for a wide variety of systems and tasks.

#### Other info (issues closed, discussion etc)
The unit tests aren't in yet - this one's going to be a doozy to test, I think - but it's ready to get some eyes on it! After this, I'm going to make one more expansion (items and status effects) before combining them all into a complete system.

UPDATE: I've merged in my items/status effects system. This one isn't quite ready to go yet, as the documentation and examples aren't done yet, but most of the base functionality is in and working properly.

chainsol commented:
I am super excited about this! Ainneve has had an issue open for a long time about adding a magic system, and I'm hoping we can replace Ainneve's combat system with Turnbattle! I'm gonna read over this, and see if I have any questions! :D

FlutterSprite commented:
Since my system for items and conditions is almost ready to go too, I decided to merge that in with this one! I'll update the PR's description to reflect this.

FlutterSprite commented:
All right, this is ready to get some eyes on it, I think! I'll get started on the unit tests once everything looks good!

Griatch commented on evennia/contrib/turnbattle/tb_items.py:
```diff
@@ -0,0 +1,1374 @@
+    attack_value = randint(1, 100)
+    # Add 25 to the roll if the attacker has the "Accuracy Up" condition.
+    if "Accuracy Up" in attacker.db.conditions:
+        attack_value += 25
```
I know these values are here partly as examples, but these kind of numbers seem to be perfect to have up top as `ACCURACY_UP_STEP` etc. Not only does it move the data values to one place where they are easy to change, it also serves as a good example for people expanding on the system (the more these numbers are spread out in the module the easier it is to miss updating one when tweaking). 

This should be considered for the other cases below too, I won't point those out explicitly.

FlutterSprite replied:
That's a good point! I'll add condition values to the options.

Griatch commented on evennia/contrib/turnbattle/tb_items.py:
```diff
@@ -0,0 +1,1374 @@
+    # For this example, just return 50, for about a 50/50 chance of hit.
+    defense_value = 50
+    # Add 15 to defense if the defender has the "Defense Up" condition.
+    if "Defense Up" in defender.db.conditions:
```
Out of curiosity, why does the `Defense Up` Attribute not itself store the value `15`? That way a given defender could have different defense value from another. It sounds quite logical to store such things with the object.

FlutterSprite replied:
I thought about this, but decided that having different instances of the same conditions with varying levels of severity is an exercise best left to the developer, since different rules systems handle it so differently.

Griatch commented on evennia/contrib/turnbattle/tb_items.py:
```diff
@@ -0,0 +1,1374 @@
+
+"""
+----------------------------------------------------------------------------
+PROTOTYPES START HERE
```
Cool way to offer some starter objects to use!

Griatch commented on evennia/contrib/turnbattle/tb_magic.py:
```diff
@@ -0,0 +1,1285 @@
+        character (obj): Character to have their combat attributes removed
+
+    Notes:
+        Any attribute whose key begins with 'combat_' is temporary and no
```
Attributes starting with `combat_` does not seem too unlikely. Since this is not warned for, this could cause some unfortunate bugs and weirdness if people just drop this into their existing code.  A safer (and more efficient) way if you want to use temporary Attributes like this is to store the `Attribute` with a `category` set to `combat`. It can then be named whatever you like, and you later just delete all attributes of the `combat` category: 

```
obj.attributes.add("Burning bonus", 1000, category="combat")
obj.attributes.add("Disadvantage", -100 , category="combat")
# removing all of a given category
obj.attributes.clear(category="combat") 
```

The even cleaner solution would be to store all such temporary variables on the Combat-Handler Script instead, indexed with who they belong to. That way they will automatically go away when the combat ends and there is never any risk of any leftovers.

FlutterSprite replied:
I had no idea attribute categories were a thing! Thats probably a better way to do it, but it may be a while before I can integrate it properly, since it'll be a big change in every module. I'll save this for another PR.

FlutterSprite commented:
Thanks for the feedback! Regarding having the modules be more modular and mixable - one concern I have with that is that none of these systems are designed to be plopped into a game and used as-is, but instead meant as a base to work from for developers to implement their own rules systems - and I worry that spending a lot of effort making the code really modular would make it harder to work with in this regard. Plus, if developers want to make these systems interoperate, they'd need to splice them together manually anyhow. I admit that another part of it is that I'm still a pretty novice coder, so things like mixins and class hierarchy still don't come terribly natural to me! That said, I agree that having swaths of identical code in each module is pretty inelegant, and I'm definitely open to refactoring how this all goes together in the future, but that may be a project for another time.

My next project for this system is to make a "complete" version that implements all these systems as well as interactions between them - maybe the best thing to do is to have the "basic" and "complete" modules implement the whole system, while the other modules just add their systems onto "basic" instead of repeating any code.

chainsol commented:
I will say, since I plan on implementing this in Ainneve, I imagine we'll be submitting patches for it.

FlutterSprite commented:
The unit tests work just fine on my local machine, but for some reason this fails the Travis CI build and I'm just completely stumped as to why! Can anyone help out...?

FlutterSprite commented:
What the heck - I commented out the parts that seemed like they were a problem, but it's still failing...!

FlutterSprite commented:
I just commented out the entire section for unit tests for tb_items. I'll get started on tests for tb_magic soon and come back to this later.

FlutterSprite commented:
I did it! I found a way to make everything build nice - I just made another version of the character that didn't use the TickerHandler, and it was smooth sailing from there! I'll finish the rest of the unit tests soon!

FlutterSprite commented:
All right, unit tests for the magic system stuff are in too!

Griatch commented:
@FlutterSprite Good job! I'll look this over soon, but it may take a few days (these PRs are pretty big after all!)

Griatch commented on evennia/contrib/tests.py:
```diff
@@ -1214,6 +1238,199 @@ def test_tbrangefunc(self):
+        user.db.conditions = {"Poisoned":[5, user]}
+        tb_items.itemfunc_cure_condition(test_healpotion, user, user)
+        self.assertTrue(user.db.conditions == {})
+        # Delete the test character
```
This kind of cleanup belongs in a `tearDown` method, it will then be called after automatically after each test-method in the class.

Griatch commented on evennia/contrib/turnbattle/tb_items.py:
```diff
@@ -0,0 +1,1397 @@
+----------------------------------------------------------------------------
+"""
+
+TURN_TIMEOUT = 30 # Time before turns automatically end, in seconds
```
The constants up here makes it much cleaner. :)

Griatch commented on evennia/contrib/turnbattle/tb_items.py:
```diff
@@ -0,0 +1,1397 @@
+        self.db.conditions = {} # Set empty dict for conditions
+        # Subscribe character to the ticker handler
+        tickerhandler.add(NONCOMBAT_TURN_TIME, self.at_update, idstring="update")
+        """
```
This docstring belongs with the other docstring at the top of the method. You can make a `Notes:` section if you want to differentiate it from the header. Since docstrings are generally only used for methods/functions/classes, having it down here looks like a method went missing ... If you want a longer comment later in a method, use `#` comments instead.

Griatch commented on evennia/contrib/turnbattle/tb_items.py:
```diff
@@ -0,0 +1,1397 @@
+
+These functions carry out the action of using an item - every item should
+contain a db entry "item_func", with its value being a string that is
+matched to one of these functions in the ITEMFUNCS dictionary below.
```
While this way of mapping is fine, a more Evennia-esque way would be to have these functions in a separate module and use the `callables_from_module` Evennia utility - it returns a dictionary keyed by their `def` names in the module, with values being the callables. That way you could easily have people add new stuff by just having them add the path to a new module with item-funcs to some setting (see for example how lockfuncs, inlinefuncs etc work). Not necessary but probably more consistent with how other similar function-plugins work in Evennia.

FlutterSprite replied:
Seems good! I'll keep that in mind - I'm interested in going through all the turnbattle modules and kind of cleaning them up a bit at some point.

Griatch commented on evennia/contrib/turnbattle/tb_magic.py:
```diff
@@ -0,0 +1,1293 @@
+
+This module is meant to be heavily expanded on, so you may want to copy it
+to your game's 'world' folder and modify it there rather than importing it
+in your game and using it as-is.
```
Good docstring here!

Griatch commented on evennia/contrib/turnbattle/tb_magic.py:
```diff
@@ -0,0 +1,1293 @@
+        self.db.spells_known = [] # Set empty spells known list
+        self.db.max_mp = 20 # Set maximum MP to 20
+        self.db.mp = self.db.max_mp # Set current MP to maximum
+        """
```
Please move this up with the normal docstring of this method. Use a `Notes:` section if you want to separate it from the header.

Griatch commented on evennia/contrib/turnbattle/tb_magic.py:
```diff
@@ -0,0 +1,1293 @@
+            spelldata["spellfunc"](caller, spell_to_cast, spell_targets, spelldata["cost"], **kwargs)
+        except Exception:
+            log_trace("Error in callback for spell: %s." % spell_to_cast)
+        """
```
This should be moved to the end of the main docstring or made into a comment with `#`.

Griatch commented on evennia/contrib/turnbattle/tb_magic.py:
```diff
@@ -0,0 +1,1293 @@
+        self.caller.db.mp = self.caller.db.max_mp  # Set current MP to maximum
+        self.caller.location.msg_contents("%s rests to recover HP and MP." % self.caller)
+        """
+        You'll probably want to replace this with your own system for recovering HP and MP.
```
Since this type of commenting is inconstence with anything else in Evennia, please make this into a comment with `#` or move it together with the regular doc string.

Griatch commented:
@FlutterSprite Any further progress on this?

FlutterSprite commented:
Sorry! I've been really busy with my main project (which just got announced! Check it out! http://www.grappleforce.com ) so I haven't really had any time to go back to Evennia stuff lately. Once things slow down for me a bit, I'll get back to getting this squared away.

Griatch commented:
@FlutterSprite Oh, I remember you showing me some animations from this before! That grappling-hook thingie remains very cool and I like the artwork and details of this. Very cool. :) Glad things are progressing!

Griatch commented:
@FlutterSprite Gotten any more time to wrap this one up?

FlutterSprite commented:
Sorry! My main project still is eating up most of my time and attention. I re-read through your comments and I may be able to make some of the more minor tweaks and syntax things soon so we can get everything merged - I don't think I'll have any time for any major overhauls to how this is all put together (especially the stuff with unit tests, which I still have only the barest familiarity with). If there's anyone else interested in helping get it finished up, I'd really appreciate it!

Griatch commented:
@FlutterSprite Bump.

FlutterSprite commented:
I'm probably going to take a few days or a week off from my main project sometime soon, so I'll get back to this during that time and hopefully get it closed out!

FlutterSprite commented on evennia/contrib/tests.py:
```diff
@@ -962,6 +962,30 @@ def test_turnbattlerangecmd(self):
+        
+    # Test magic commands
+    def test_turnbattlemagiccmd(self):
+        self.call(tb_magic.CmdStatus(), "", "You have 100 / 100 HP and 20 / 20 MP.")
```
Currently, yes. This is how it is for all the 'turnbattle' modules - I think going through and cleaning them up and making the tests more thorough and consistent might be its own project, part of making all the 'turnbattle' modules more streamlined.

FlutterSprite commented:
Hoping to finally resolve all this stuff today!

FlutterSprite commented:
All right, I split the Turnbattle modules' unit test classes up more, and refactored stuff into setUp and tearDown methods, so it should all be a lot cleaner and more proper now. This should address everything brought up before. I really hope this is enough to get it all merged in!

Griatch commented:
There are a few comments that needs to be kept in mind, but I don't think there is a need to hold off on merging this. :) 

The main issue I have with this as it stands however is that it appears the unit tests are not isolated - probably due to the tickerhandler components of these tests (dirty reactor), it clashes with other tests and fails the test suite.

Griatch commented:
There - patched out the tickerhandler in the unit test. Merged now :)
